File: integrations/kagent/appa-kagent-adk-go/mcp_lifecycle.go
package appakagentadk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"slices"
	"sort"
	"strings"
	"sync"

	kagentagent "github.com/kagent-dev/kagent/go/adk/pkg/agent"
	"github.com/kagent-dev/kagent/go/adk/pkg/sts"
	"github.com/kagent-dev/kagent/go/api/adk"
	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/model"
	"google.golang.org/adk/v2/tool"
	"google.golang.org/adk/v2/tool/toolutils"
	"google.golang.org/genai"
)

type observedTool struct {
	Name string `json:"name"`
	Tool string `json:"tool"`
}
type inventorySource struct {
	Server  string          `json:"server"`
	Status  DiscoveryStatus `json:"status"`
	Dynamic bool            `json:"dynamic"`
	Detail  string          `json:"detail,omitempty"`
}
type observedInventory struct {
	Tools   []observedTool    `json:"tools"`
	Sources []inventorySource `json:"sources"`
}
type toolCheck struct {
	Tool   string `json:"tool"`
	Status string `json:"status"`
	Reason string `json:"reason"`
}
type validationReport struct {
	Tools       []toolCheck    `json:"tools"`
	Errors      []string       `json:"errors"`
	Accepted    []observedTool `json:"accepted_tools"`
	ActorOpened bool           `json:"actor_opened"`
}

type mcpConnection struct {
	params    mcpServerParams
	names     []string
	approvals map[string]bool
	server    string
	control   bool
}
type mcpRun struct {
	mu           sync.Mutex
	sets         []*discoveredToolset
	observations []toolsetObservation
	tools        [][]tool.Tool
	inventory    observedInventory
	names        Inventory
	opened       bool
	selected     map[string]*discoveredMCPTool
}

// MCPDiscovery is the invocation-owned MCP tool source installed by the gated
// runner. It is not callable and does not add a model-visible command.
type MCPDiscovery struct {
	connections []mcpConnection
	mu          sync.Mutex
	runs        map[string]*mcpRun
	plugin      *AppaPluginKagent
}

func (*MCPDiscovery) Name() string        { return "appa_mcp_tools" }
func (*MCPDiscovery) Description() string { return "" }
func (*MCPDiscovery) IsLongRunning() bool { return false }

func newMCPDiscovery(config *adk.AgentConfig, stsPlugin *sts.TokenPropagationPlugin) (*MCPDiscovery, error) {
	// The gated entrypoint appends this reserved source after checking user
	// configuration. Never interpret an arbitrary last source as runtime tools.
	if config == nil || len(config.HttpTools) == 0 {
		return nil, failClosed("gated runner requires its reserved runtime toolset")
	}
	reserved := config.HttpTools[len(config.HttpTools)-1].Tools
	if !slices.Equal(reserved, []string{ReservedTool}) && !slices.Equal(reserved, RuntimeTools) {
		return nil, failClosed("gated runner has no reserved runtime toolset")
	}
	seen := make(map[string]bool)
	check := func(endpoint string) error {
		id, err := mcpSourceID(endpoint)
		if err != nil {
			return err
		}
		if seen[id] {
			return failClosed("MCP endpoint is configured more than once; combine its tool filters and approval requirements")
		}
		seen[id] = true
		return nil
	}
	for _, c := range config.HttpTools {
		if err := check(c.Params.Url); err != nil {
			return nil, err
		}
	}
	for _, c := range config.SseTools {
		if err := check(c.Params.Url); err != nil {
			return nil, err
		}
	}
	result := &MCPDiscovery{runs: make(map[string]*mcpRun)}
	var provider DynamicHeaderProvider
	if stsPlugin != nil {
		provider = stsPlugin.HeaderProvider
	}
	propagate := strings.EqualFold(os.Getenv("KAGENT_PROPAGATE_TOKEN"), "true")
	appendConnection := func(params mcpServerParams, names, approvals []string, control bool) {
		params.PropagateToken, params.HeaderProvider = propagate, provider
		required := make(map[string]bool)
		for _, name := range approvals {
			required[name] = true
		}
		server, _ := mcpSourceID(params.URL) // Every endpoint was checked above.
		result.connections = append(result.connections, mcpConnection{params: params, names: names, approvals: required, server: server, control: control})
	}
	for index, c := range config.HttpTools {
		appendConnection(mcpServerParams{URL: c.Params.Url, Headers: c.Params.Headers, AllowedHeaders: c.AllowedHeaders, ServerType: "http", Timeout: c.Params.Timeout, SseReadTimeout: c.Params.SseReadTimeout, TLSInsecureSkipVerify: c.Params.TLSInsecureSkipVerify, TLSCACertPath: c.Params.TLSCACertPath, TLSDisableSystemCAs: c.Params.TLSDisableSystemCAs}, c.Tools, c.RequireApproval, index == len(config.HttpTools)-1)
	}
	for _, c := range config.SseTools {
		appendConnection(mcpServerParams{URL: c.Params.Url, Headers: c.Params.Headers, AllowedHeaders: c.AllowedHeaders, ServerType: "sse", Timeout: c.Params.Timeout, SseReadTimeout: c.Params.SseReadTimeout, TLSInsecureSkipVerify: c.Params.TLSInsecureSkipVerify, TLSCACertPath: c.Params.TLSCACertPath, TLSDisableSystemCAs: c.Params.TLSDisableSystemCAs}, c.Tools, c.RequireApproval, false)
	}
	return result, nil
}

func (d *MCPDiscovery) run(id string) *mcpRun {
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.runs[id] == nil {
		d.runs[id] = &mcpRun{sets: make([]*discoveredToolset, len(d.connections)), observations: make([]toolsetObservation, len(d.connections)), tools: make([][]tool.Tool, len(d.connections))}
	}
	return d.runs[id]
}

func (d *MCPDiscovery) close(id string) {
	d.mu.Lock()
	run := d.runs[id]
	delete(d.runs, id)
	d.mu.Unlock()
	if run == nil {
		return
	}
	run.mu.Lock()
	defer run.mu.Unlock()
	run.opened = false
	for _, observation := range run.observations {
		if observation.session != nil {
			observation.session.Close()
		}
	}
}

var errFamilyNotOpened = errors.New("APPA family has not opened")

func (d *MCPDiscovery) validate(ctx context.Context, ids trajectoryIDs, inventory observedInventory, pinned bool) (validationReport, error) {
	ctx, cancel := context.WithTimeout(ctx, gatedTimeout)
	defer cancel()
	request := map[string]any{"protocol": Protocol, "adapter": Adapter, "inventory": inventory}
	if pinned {
		request["root_id"] = ids.rootID
		if ids.childID != "" {
			request["child_id"] = ids.childID
		}
	}
	body, err := json.Marshal(request)
	if err != nil {
		return validationReport{}, err
	}
	req, err := http.NewRequestWithContext(ctx, "POST", strings.TrimSuffix(d.plugin.hookURL, "/hook")+"/validate", strings.NewReader(string(body)))
	if err != nil {
		return validationReport{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	response, err := d.plugin.client.Do(req)
	if err != nil {
		return validationReport{}, failClosed("MCP inventory validation is unavailable")
	}
	defer response.Body.Close()
	if pinned && response.StatusCode == http.StatusNotFound {
		return validationReport{}, errFamilyNotOpened
	}
	if response.StatusCode != http.StatusOK {
		return validationReport{}, failClosed("MCP inventory validation returned HTTP %d", response.StatusCode)
	}
	var report validationReport
	data, err := io.ReadAll(io.LimitReader(response.Body, discoveryMaxBytes+1))
	if err != nil || len(data) > discoveryMaxBytes || json.Unmarshal(data, &report) != nil || report.Tools == nil || report.Errors == nil || report.Accepted == nil {
		return report, failClosed("invalid MCP inventory validation response")
	}
	checked := make(map[string]string)
	for _, check := range report.Tools {
		if check.Tool == "" || (check.Status != "valid" && check.Status != "invalid" && check.Status != "unknown") {
			return report, failClosed("invalid MCP inventory validation status")
		}
		checked[check.Tool] = check.Status
	}
	for _, observed := range inventory.Tools {
		if observed.Tool == ControlTool {
			continue // Runtime control is not an authored policy contract.
		}
		if checked[observed.Name] != "valid" && checked[observed.Name] != "invalid" {
			return report, failClosed("MCP inventory validation omitted a known tool")
		}
	}
	return report, nil
}

// prepare is called before the opening prompt and on each model request. A
// configured source failing to enumerate is unknown; invalid new tools are
// isolated after opening. Runtime admission still reserves every accepted name.
func (d *MCPDiscovery) prepare(ctx agent.ReadonlyContext, ids trajectoryIDs, pinned bool) (*mcpRun, error) {
	run := d.run(ctx.InvocationID())
	run.mu.Lock()
	defer run.mu.Unlock()
	// A fresh host process may be continuing an existing APPA trajectory. Only
	// an explicit missing-family response permits using the serving policy.
	previous, err := d.validate(ctx, ids, observedInventory{Tools: []observedTool{}, Sources: []inventorySource{}}, true)
	if errors.Is(err, errFamilyNotOpened) && !pinned && ids.childID == "" {
		previous, err = d.validate(ctx, ids, observedInventory{Tools: []observedTool{}, Sources: []inventorySource{}}, false)
	} else if err == nil {
		pinned = true
		if previous.ActorOpened {
			run.opened = true
		}
	}
	if err != nil {
		return nil, err
	}
	limit := make(chan struct{}, 4)
	errors := make([]error, len(d.connections))
	var workers sync.WaitGroup
	for index, connection := range d.connections {
		workers.Add(1)
		go func(index int, connection mcpConnection) {
			defer workers.Done()
			select {
			case limit <- struct{}{}:
				defer func() { <-limit }()
			case <-ctx.Done():
				errors[index] = ctx.Err()
				return
			}
			if run.sets[index] == nil {
				transport, err := createTransport(ctx, connection.params)
				if err != nil {
					errors[index] = fmt.Errorf("MCP source %s has invalid transport configuration", connection.server)
					return
				}
				set, err := newDiscoveredToolset(transport, connection.names)
				if err != nil {
					errors[index] = err
					return
				}
				run.sets[index] = set
			}
			found, observation, err := run.sets[index].discover(ctx)
			if observation.session != nil {
				run.observations[index].session = observation.session
			}
			if err != nil {
				errors[index] = err
				run.observations[index].discovery = Discovery{Status: DiscoveryUnavailable, Detail: "MCP metadata update is invalid; previous tools retained"}
				return
			}
			run.observations[index] = observation
			if observation.discovery.Status != DiscoveryUnavailable {
				if observation.discovery.Status == DiscoveryPartial {
					retained := make(map[string]tool.Tool)
					for _, candidate := range run.tools[index] {
						retained[candidate.Name()] = candidate
					}
					for _, candidate := range found {
						retained[candidate.Name()] = candidate
					}
					if len(retained) > discoveryMaxTools {
						errors[index] = failClosed("MCP partial inventory exceeds its tool limit")
						return
					}
					found = make([]tool.Tool, 0, len(retained))
					for _, candidate := range retained {
						found = append(found, candidate)
					}
				}
				run.tools[index] = found
			}
		}(index, connection)
	}
	workers.Wait()
	if ctx.Err() != nil {
		return nil, ctx.Err()
	}
	if !run.opened {
		for _, err := range errors {
			if err != nil {
				return nil, err
			}
		}
	}
	for index, err := range errors {
		if err != nil {
			log.Printf("appa: MCP source %s metadata update refused; previous tools retained", d.connections[index].server)
		}
	}
	inventory := observedInventory{Tools: []observedTool{}, Sources: []inventorySource{}}
	for name, spelling := range d.plugin.inventory.spellings {
		inventory.Tools = append(inventory.Tools, observedTool{Name: name, Tool: spelling})
	}
	candidates := make(map[string][]*discoveredMCPTool)
	for index, connection := range d.connections {
		observation := run.observations[index].discovery
		if observation.Status == "" {
			observation.Status = DiscoveryUnavailable
		}
		if !connection.control {
			inventory.Sources = append(inventory.Sources, inventorySource{Server: connection.server, Status: observation.Status, Dynamic: true, Detail: observation.Detail})
		}
		for _, candidate := range run.tools[index] {
			spelling := MCPSpelling(connection.server, candidate.Name())
			if connection.control {
				var ok bool
				spelling, ok = d.plugin.inventory.Spelling(candidate.Name())
				if !ok {
					return nil, failClosed("unregistered runtime control tool")
				}
			}
			runnable, ok := candidate.(mcpRunnable)
			if !ok {
				return nil, failClosed("MCP tool has no executable declaration")
			}
			candidates[candidate.Name()] = append(candidates[candidate.Name()], &discoveredMCPTool{mcpRunnable: runnable, spelling: spelling, approvals: connection.approvals})
		}
	}
	// Fetch durable reservations first: a restarted plugin cannot let a new
	// source claim a name already owned by another source in this actor.
	accepted := make(map[string]string)
	for _, tool := range previous.Accepted {
		accepted[tool.Name] = tool.Tool
	}
	selected := make(map[string]*discoveredMCPTool)
	for name, choices := range candidates {
		static, isStatic := d.plugin.inventory.Spelling(name)
		for _, candidate := range choices {
			if isStatic && candidate.spelling != static {
				continue
			}
			if old, exists := accepted[name]; exists && old != candidate.spelling {
				continue
			}
			if selected[name] != nil {
				if !run.opened {
					return nil, failClosed("MCP tool name %s is ambiguous", name)
				}
				delete(selected, name)
				break
			}
			selected[name] = candidate
		}
		if !run.opened && selected[name] == nil {
			return nil, failClosed("MCP tool name %s conflicts with an existing identity", name)
		}
	}
	for name, candidate := range selected {
		if _, static := d.plugin.inventory.Spelling(name); !static {
			inventory.Tools = append(inventory.Tools, observedTool{Name: name, Tool: candidate.spelling})
		}
	}
	sort.Slice(inventory.Tools, func(i, j int) bool { return inventory.Tools[i].Name < inventory.Tools[j].Name })
	report, err := d.validate(ctx, ids, inventory, pinned)
	if err != nil {
		return nil, err
	}
	if len(report.Errors) > 0 {
		return nil, failClosed("MCP inventory is invalid: %s", strings.Join(report.Errors, "; "))
	}
	invalid := make(map[string]bool)
	for _, check := range report.Tools {
		if check.Status == "invalid" {
			// Missing coverage disables this tool, not the conversation.
			invalid[check.Tool] = true
			delete(selected, check.Tool)
		}
	}
	filtered := inventory.Tools[:0]
	run.names = Inventory{spellings: make(map[string]string), names: make(map[string]string)}
	for _, tool := range inventory.Tools {
		if !invalid[tool.Name] {
			filtered = append(filtered, tool)
			run.names.spellings[tool.Name] = tool.Tool
			run.names.names[tool.Tool] = tool.Name
		}
	}
	inventory.Tools = filtered
	run.inventory = inventory
	run.selected = selected
	return run, nil
}

type mcpRunnable interface {
	tool.Tool
	Declaration() *genai.FunctionDeclaration
	Run(agent.Context, any) (map[string]any, error)
}
type discoveredMCPTool struct {
	mcpRunnable
	spelling  string
	approvals map[string]bool
}

func (t *discoveredMCPTool) ProcessRequest(ctx agent.Context, req *model.LLMRequest) error {
	return toolutils.PackTool(req, t)
}
func (t *discoveredMCPTool) Run(ctx agent.Context, args any) (map[string]any, error) {
	values, ok := args.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("MCP arguments must be an object")
	}
	result, err := kagentagent.MakeApprovalCallback(t.approvals)(ctx, t, values)
	if result != nil || err != nil {
		return result, err
	}
	return t.mcpRunnable.Run(ctx, args)
}

func (d *MCPDiscovery) ProcessRequest(ctx agent.Context, req *model.LLMRequest) error {
	if d.plugin == nil {
		return failClosed("MCP discovery is not attached to the APPA plugin")
	}
	ids, ok := d.plugin.idsFor(ctx)
	if !ok {
		return failClosed("MCP discovery has no trajectory scope")
	}
	run, err := d.prepare(ctx, ids, true)
	if err != nil {
		return err
	}
	run.mu.Lock()
	defer run.mu.Unlock()
	names := make([]string, 0, len(run.selected))
	for name := range run.selected {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		if err := toolutils.PackTool(req, run.selected[name]); err != nil {
			return err
		}
	}
	apps := make(map[string]bool)
	for _, observation := range run.observations {
		for _, candidate := range observation.discovery.Tools {
			ui, _ := candidate.Meta["ui"].(map[string]any)
			uri, _ := ui["resourceUri"].(string)
			if uri == "" {
				uri, _ = candidate.Meta["ui/resourceUri"].(string)
			}
			if uri != "" && run.selected[candidate.Name] != nil {
				apps[candidate.Name] = true
			}
		}
	}
	if len(apps) > 0 {
		_, err := kagentagent.MakeMCPAppModelResultCallback(apps)(ctx, req)
		if err != nil {
			return err
		}
	}
	return nil
}

File: integrations/kagent/appa-kagent-adk/src/appa_kagent_adk/mcp_lifecycle.py
"""Invocation-scoped MCP evidence, using kagent's existing authenticated tools.

The host owns transports, credentials, approvals and execution. This toolset
enumerates metadata and checks policy coverage before exposing callable tools.
"""

from __future__ import annotations

import asyncio
import logging
import weakref
from dataclasses import dataclass, field

from google.adk.tools.base_toolset import BaseToolset
from google.adk.tools.mcp_tool.mcp_tool import McpTool

from . import wire
from .config_guard import ConfigRefused
from .discovery import MAX_BYTES, MAX_TOOLS, Discovery, discover_toolset
from .inventory import ToolInventory, mcp_source_id, mcp_spelling

logger = logging.getLogger(__name__)


@dataclass
class _Run:
    lock: asyncio.Lock = field(default_factory=asyncio.Lock)
    opened: bool = False
    evidence: dict = field(default_factory=lambda: {"tools": [], "sources": []})
    selected: list = field(default_factory=list)
    # Keep identities for tool handles still held by the host, without retaining
    # every wrapper created by earlier model requests for the whole invocation.
    identities: weakref.WeakKeyDictionary = field(default_factory=weakref.WeakKeyDictionary)
    observations: list = field(default_factory=list)
    source_tools: list = field(default_factory=list)
    names: ToolInventory = field(default_factory=lambda: ToolInventory({}))


class MCPDiscovery(BaseToolset):
    def __init__(self, sources, plugin):
        super().__init__()
        self._use_invocation_cache = False
        self.sources = tuple(sources)
        self.plugin = plugin
        self.runs: dict[str, _Run] = {}
        self.servers = tuple(mcp_source_id(source._connection_params.url) for source in sources)
        if len(set(self.servers)) != len(self.servers):
            raise ConfigRefused("MCP endpoint is configured more than once; combine its filters and approvals")
        for source in sources:
            if source.tool_name_prefix or source._use_mcp_resources:
                raise ConfigRefused("MCP discovery requires native tool names and an explicit resource-tool identity")

    def state(self, invocation_id):
        if invocation_id not in self.runs:
            self.runs[invocation_id] = _Run(
                observations=[Discovery((), "unavailable") for _ in self.sources],
                source_tools=[[] for _ in self.sources],
            )
        return self.runs[invocation_id]

    async def _validate(self, root, child, inventory, *, pinned):
        try:
            return await asyncio.wait_for(self._request_validation(root, child, inventory, pinned=pinned), timeout=120)
        except asyncio.TimeoutError:
            raise ConfigRefused("MCP inventory validation timed out") from None

    async def _request_validation(self, root, child, inventory, *, pinned):
        request = {"protocol": wire.PROTOCOL, "adapter": wire.ADAPTER, "inventory": inventory}
        if pinned:
            request["root_id"] = root
            if child is not None:
                request["child_id"] = child
        try:
            async with self.plugin._live_client().stream(
                "POST", self.plugin._hook_url.removesuffix("/hook") + "/validate", json=request
            ) as response:
                if response.status_code == 404 and pinned:
                    return None
                if response.status_code != 200:
                    raise ConfigRefused(f"MCP inventory validation returned HTTP {response.status_code}")
                data = bytearray()
                async for chunk in response.aiter_bytes():
                    data.extend(chunk)
                    if len(data) > MAX_BYTES:
                        raise ConfigRefused("MCP validation response exceeds its size limit")
            import json

            report = json.loads(data)
        except ConfigRefused:
            raise
        except Exception:
            raise ConfigRefused("MCP inventory validation is unavailable or malformed") from None
        if not isinstance(report, dict) or any(
            not isinstance(report.get(key), list) for key in ("tools", "errors", "accepted_tools")
        ):
            raise ConfigRefused("MCP inventory validation response is incomplete")
        if not isinstance(report.get("actor_opened", False), bool):
            raise ConfigRefused("MCP inventory validation contains an invalid actor state")
        checked = {}
        for check in report["tools"]:
            if (
                not isinstance(check, dict)
                or not isinstance(check.get("tool"), str)
                or check.get("status") not in ("valid", "invalid", "unknown")
            ):
                raise ConfigRefused("MCP inventory validation contains an invalid status")
            checked[check["tool"]] = check["status"]
        for observed in inventory["tools"]:
            if observed["tool"] != wire.CONTROL_TOOL and checked.get(observed["name"]) not in ("valid", "invalid"):
                raise ConfigRefused("MCP inventory validation omitted a known tool")
        return report

    def _convert(self, source, metadata):
        # Same constructor fields in pinned ADK 1.31.1 and 2.8.0. Keep the
        # original session manager, auth, confirmation and progress handling.
        candidate = McpTool(
            mcp_tool=metadata,
            mcp_session_manager=source._mcp_session_manager,
            auth_scheme=source._auth_scheme,
            auth_credential=source._auth_credential,
            require_confirmation=source._require_confirmation,
            header_provider=source._header_provider,
            progress_callback=getattr(source, "_progress_callback", None),
        )
        from kagent.adk._mcp_toolset import ConnectionSafeMcpTool

        return ConnectionSafeMcpTool(candidate)

    async def prepare(self, context, *, allow_new=False):
        invocation = context.invocation_id
        self.plugin._discovery_runs[invocation] = self
        state = self.state(invocation)
        root, child = self.plugin._ids(context)
        async with state.lock:
            empty = {"tools": [], "sources": []}
            previous = await self._validate(root, child, empty, pinned=True)
            pinned = previous is not None
            if not pinned:
                if not allow_new or child is not None:
                    raise ConfigRefused("the requested APPA family has not opened")
                previous = await self._validate(root, child, empty, pinned=False)
            elif previous.get("actor_opened", False):
                state.opened = True
            accepted = {}
            for observed in previous["accepted_tools"]:
                if not isinstance(observed, dict) or not all(
                    isinstance(observed.get(key), str) for key in ("name", "tool")
                ):
                    raise ConfigRefused("MCP validation returned an invalid accepted identity")
                accepted[observed["name"]] = observed["tool"]
            semaphore = asyncio.Semaphore(4)

            async def inspect(index, source):
                async with semaphore:
                    try:
                        observed = await discover_toolset(source, context)
                        if observed.status != "unavailable":
                            try:
                                candidates = [self._convert(source, metadata) for metadata in observed.tools]
                            except ValueError:
                                raise ConfigRefused("MCP tool metadata cannot be converted by the host") from None
                            if observed.status == "partial":
                                retained = {tool.name: tool for tool in state.source_tools[index]}
                                retained.update({tool.name: tool for tool in candidates})
                                if len(retained) > MAX_TOOLS:
                                    raise ConfigRefused("MCP partial inventory exceeds its tool limit")
                                candidates = list(retained.values())
                            state.source_tools[index] = candidates
                        state.observations[index] = observed
                    except ConfigRefused:
                        if not state.opened:
                            raise
                        state.observations[index] = Discovery(
                            (), "unavailable", "invalid MCP metadata update; previous tools retained"
                        )
                        logger.warning(
                            "MCP source %s metadata update refused; previous tools retained", self.servers[index]
                        )

            results = await asyncio.gather(
                *(inspect(i, source) for i, source in enumerate(self.sources)), return_exceptions=True
            )
            for result in results:
                if isinstance(result, BaseException):
                    raise result
            base = self.plugin._inventory
            choices = {}
            for server, tools in zip(self.servers, state.source_tools):
                for candidate in tools:
                    choices.setdefault(candidate.name, []).append((candidate, mcp_spelling(server, candidate.name)))
            selected = {}
            for name, candidates in choices.items():
                eligible = [
                    (tool, identity)
                    for tool, identity in candidates
                    if name not in base.spellings and accepted.get(name, identity) == identity
                ]
                if len(eligible) != 1:
                    if not state.opened:
                        raise ConfigRefused(
                            f"MCP tool name {name!r} is ambiguous or conflicts with an existing identity"
                        )
                    continue
                selected[name] = eligible[0]
            names = dict(base.spellings)
            names.update({name: identity for name, (_, identity) in selected.items()})
            inventory = {
                "tools": [{"name": name, "tool": identity} for name, identity in sorted(names.items())],
                "sources": [
                    {
                        "server": server,
                        "status": observed.status,
                        "dynamic": True,
                        **({"detail": observed.detail} if observed.detail else {}),
                    }
                    for server, observed in zip(self.servers, state.observations)
                ],
            }
            report = await self._validate(root, child, inventory, pinned=pinned)
            if report is None or report["errors"]:
                raise ConfigRefused("MCP inventory configuration is invalid")
            invalid = {check["tool"] for check in report["tools"] if check["status"] == "invalid"}
            # Missing policy coverage disables a tool, not the other tools in
            # this conversation. Only covered identities enter the gated inventory.
            inventory["tools"] = [entry for entry in inventory["tools"] if entry["name"] not in invalid]
            state.names = ToolInventory({entry["name"]: entry["tool"] for entry in inventory["tools"]})
            state.selected = [tool for name, (tool, _) in sorted(selected.items()) if name not in invalid]
            for name, (tool, identity) in selected.items():
                if name not in invalid:
                    state.identities[tool] = identity
            state.evidence = inventory
            return state

    async def get_tools(self, readonly_context=None):
        if readonly_context is None:
            raise ConfigRefused("MCP discovery requires an invocation scope")
        return (await self.prepare(readonly_context)).selected

    async def close(self):
        try:
            results = await asyncio.gather(*(source.close() for source in self.sources), return_exceptions=True)
            for result in results:
                if isinstance(result, BaseException):
                    raise result
        finally:
            for invocation in self.runs:
                if self.plugin._discovery_runs.get(invocation) is self:
                    self.plugin._discovery_runs.pop(invocation)
            self.runs.clear()
